### DISCUSSION TITLE
Noob question - where (on webpage) are plugins located? Have a few installed, can't find any of them

### DISCUSSION THREAD

**Discussion by treimers1892** [netbox-community/netbox#11103] [Q&A]
Hi all --

Maybe I'm having a seriously blind spot --

But I can't actually figure out where the "Plugins" link is in the main webpage for Netbox.

Main webpages work - I've added an Org, site, racks, devices, etc.

Running Netbox 3.3.8 loaded in a VM of Centos 8 Stream, fully updated

Specifically, I'm trying to use NetDoc by Andrea Dainese
I've followed the directions here.
https://github.com/dainok/netdoc

I have followed the directions here for the basics, as well as the specific instructions from Andrea. 
https://docs.netbox.dev/en/stable/plugins/#plugins

Everything seemed to go OK - NO errors to speak of.
I even ran "update.sh" inside the Venv again.

I have other plugins installed as well, they don't appear either.
From the below, I should be seeing 
NetDoc
NetDoc _topolopgy_views
and
NetBox-UI-plugin

capabilities somewhere in the Netbox webpage.

But no Plugin link anywhere on the lefthand column of NetBox as you might expect.
I don't know where installation logs are stored that I could go look and see if there were errors that went by too fast. 

Have I done something stupid like completely fail to  "switch on" the use of Plugins in some more global fashion? 

The PLUGINS section from configuration.py

PLUGINS = [
'netdoc',
'netbox_topology_views',
'nextbox_ui_plugin',
]
PLUGINS_CONFIG = {
    'netdoc': {
        'NTC_TEMPLATES_DIR': '/opt/ntc-templates/ntc_templates/templates'
    },
    'netbox_topology_views': {
        'allow_coordinates_saving': True,
        'draw_default_layout': True,
        'draw_interface_name': True,
    }
}

I tried running "find_packages" as noted in another thread.

Bash says "not found" --- so I imagine that's some coder tool + code that smarter people than me know how to run in some Dev cli mode,
not just the bash prompt.

**Comment by treimers1892** [comment]
Hi all - 

Did I ask this in an offensive way, or fail to do something before my first post?

If there's Netbox "user" documentation that covers this, I'm unaware of where it is.
All the Google results on "netbox plugin" are the "how to install a plugin" directions, not "where to find in the webpage and how to use a plugin" directions.

Thanks

**Comment by arthanson** [comment]
If you have the plugin installed correctly, and if it does have a menu it will appear at the bottom of the left-hand side menu-bar (see screenshot):
![Monosnap Home | NetBox 2022-12-06 07-31-51](https://user-images.githubusercontent.com/99642/205954844-85256815-0c83-4167-80a1-76ee6602d7eb.png)
Note: I haven't installed the NetDoc plugin myself so not sure if it does have a menu.

**Reply by treimers1892** [reply]
Thanks Art(Hansen)

That confirms what I was thinking -- that's just about where I thought the Plugins should appear.

Now the question is why isn't that working?

Can you (or anyone else) direct me to some logging for Netbox that might shed some light on this?

I believe Django is involved somehow in plugins, is it not? Perhaps Django needs rights/permissions
isn't working correctly or something.

A bit more guidance on which corner to start looking in, and I can sort this probably. As I thought, it's not just any one plugin not working, it's the entire plugin structure that has some (possibly simple) misconfiguration somewhere.

Glad to post any parts of configs if anyone can direct me to what to post.
I believe I already did post the Plugins section from configuration.py, but here it is again:
PLUGINS = [
'netdoc',
'netbox_topology_views',
'nextbox_ui_plugin',
]
PLUGINS_CONFIG = {
    'netdoc': {
        'NTC_TEMPLATES_DIR': '/opt/ntc-templates/ntc_templates/templates'
    },
    'netbox_topology_views': {
        'allow_coordinates_saving': True,
        'draw_default_layout': True,
        'draw_interface_name': True,
    }
}

Any log files I can go looking for?

Thanks Tim

**Reply by treimers1892** [reply]
Still struggling with this.

Can anyone suggest a plug-in that does have web pages that I can try to load?

Can anyone suggest places to look to figure out why plug-ins won't appear on the menu?

**Reply by treimers1892** [reply]
Hi all -

Bumping this one up to the top again.

I still cannot figure out why the Plugins link _will not_ appear on the main lefthand menu of NetBox.
I've now tried loading _several_ different plugins, including some I'm sure have webpages.

I've run "python3 manage.py collectstatic" on each of the plugins, after adding them to the Plugins section of configuration.py.

Restarted netbox and netbox-rq several times.

Getting no errors when reinstalling the venv, with 
sudo PYTHON=/usr/bin/python3.8 ./upgrade.sh
Don't see any errors in that, and do see the upgrade script going and fetching the new plugins. 

I also run the migrate option to manage,py

I get the feeling that there's something simple wrong with permissions on a single top-level directory,
or some missing line of configuration related to django, or something.

Can anyone help?

NetBox looks like a really great platform, but I cannot convince management to use it if I cannot make it more automated for discovery
and so on.

Thanks Tim

**Reply by candlerb** [reply]
> Can anyone suggest a plug-in that does have web pages that I can try to load?

netbox-topology-views does.

> NetBox looks like a really great platform, but I cannot convince management to use it if I cannot make it more automated for discovery and so on.

That's not really how Netbox is intended to be used. See https://docs.netbox.dev/en/stable/introduction/#serve-as-a-source-of-truth

**Comment by dreng** [comment]
Hi Tim,

It is very difficult to help if one says "I followed the directions here". It is obvious that you made a mistake because the plugin system is indeed working in general and the docs are correct. As there is no obvious mistake, how should someone find out the mistake without having access to your system or looking over your shoulder?

Some thoughts:
1. Start with only one plugin in order to minimize potential mistakes.
2a. Start with a plugin that is widely used and still being maintained. I suggest netbox-topology-views.
2b. Use these instructions for netbox-topology-views as they are definitely correct: https://github.com/mattieserver/netbox-topology-views/blob/master/README.md#install
3. Login as admin and/or double check permissions for the specific menu entry in the administration portal (click your name in the upper right corner => Admin => permissions)
4. Make sure, your plugin is listed in the administration portal under "Installed Plugins".

If nothing helps, try a fresh system (Debian / Ubuntu / CentOS) and use the offical installation documentation: https://docs.netbox.dev/en/stable/installation/

**Reply by treimers1892** [reply]
Thanks, I'll start looking at that..

These were the installation directions I followed, sorry, must have lost my train of thought and not added the URL when I posted that sentence! 
https://docs.netbox.dev/en/stable/installation/

BTW - This is Netbox 3.8.8 running on Centos Stream 8

And - I've done two installs. One at work in a VM there, and one in a VM here at home.
Centos Stream 8 for both (since that's our supported platform for Linux @ work, I can't just switch to some other distro)

I'll add Topology Views and use exactly those directions, and will (probably first) check hte permissions 

Thanks Tim

**Reply by treimers1892** [reply]
OK -- update here..

In Netbox Administration under Installed Plugins, there are none showing.
So that's at least something --- the system, in spite of what's in configuration.py, and local-requirements.txt 
simply does not recognize any plugins to exist.

The netbox-topology-views site says you can just visit the URL:

In my case, thats
https://netbox.(mydomain)/plugins/netbox_topology_views/
correct?

That gives a Page Not Found.

I'm not sure what the actual DocumentRoot is for Netbox.
But I can't find a Plugins directory anywhere under /opt/netbox or below.

So something's not created right -- and downloading the plugins is putting their files where exactly?





This was the result of my following the instructions here precisely

This is what's in configuration.py relative to Plugins

# Testing just one plugin
PLUGINS = ["netbox_topology_views"]

PLUGINS_CONFIG = {
    'netbox_topology_views': {
        'preselected_device_roles': ['Router', 'Firewall'],
        'enable_circuit_terminations': True
    }
}





It looks like I had it installed already -- a "-- upgrade" was required to make it do anything.


(venv) [root@netbox netbox]# pip install netbox-topology-views --upgrade
Requirement already satisfied: netbox-topology-views in /opt/netbox-3.3.8/venv/lib/python3.8/site-packages (3.0.1)
Collecting netbox-topology-views
  Downloading netbox_topology_views-3.1.1-py3-none-any.whl (358 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 358.7/358.7 kB 1.3 MB/s eta 0:00:00
Installing collected packages: netbox-topology-views
  Attempting uninstall: netbox-topology-views
    Found existing installation: netbox-topology-views 3.0.1
    Uninstalling netbox-topology-views-3.0.1:
      Successfully uninstalled netbox-topology-views-3.0.1
Successfully installed netbox-topology-views-3.1.1
(venv) [root@netbox netbox]#
(venv) [root@netbox netbox]#
(venv) [root@netbox netbox]# python3 manage.py migrate netbox_topology_views
CommandError: No installed app with label 'netbox_topology_views'.
(venv) [root@netbox netbox]# python3 manage.py collectstatic --no-input

0 static files copied to '/opt/netbox-3.3.8/netbox/static', 520 unmodified.
(venv) [root@netbox netbox]#
(venv) [root@netbox netbox]#
(venv) [root@netbox netbox]# systemctl restart netbox netbox-rq

**Comment by treimers1892** [comment]
So - yes - the plugins dir does not exist.

I'm not sure where that goes in the various levels of "netbox"




[root@netbox netbox]# cat /var/log/netbox/netbox.log
2022-12-08 01:45:25,542 netbox.auth.logout INFO: User admin has logged out
2022-12-08 01:45:35,150 netbox.auth.login INFO: User admin successfully authenticated
2022-12-08 01:48:02,215 django.request WARNING: Not Found: /plugins
2022-12-22 17:20:37,915 netbox.auth.login INFO: User admin successfully authenticated
2022-12-23 17:10:53,355 django.request WARNING: Not Found: /favicon.ico
2022-12-23 17:18:39,706 django.request WARNING: Not Found: /favicon.ico
2022-12-23 17:29:09,362 django.request WARNING: Not Found: /plugins/netbox_topology_views/
2022-12-23 17:33:24,367 django.request WARNING: Not Found: /plugins/netbox_topology_views/
2022-12-23 17:33:36,130 django.request WARNING: Not Found: /plugins/test.html
2022-12-23 23:10:08,382 django.request WARNING: Not Found: /favicon.ico
[root@netbox netbox]#

**Comment by markkuleinio** [comment]
I have never installed NetBox plugins before but gave this plugin a shot. Install worked just fine on NetBox 3.3.8 with these instructions: https://github.com/mattieserver/netbox-topology-views/blob/master/README.md#install

- set `PLUGINS = ["netbox_topology_views"]` in `configuration.py` (replaced the default `PLUGINS = []` line)
- `source /opt/netbox/venv/bin/activate`
- `cd /opt/netbox/netbox`
- `pip3 install netbox-topology-views`
- `python3 manage.py migrate netbox_topology_views`
- `python3 manage.py collectstatic --no-input`
- `systemctl restart netbox netbox-rq`

```
(venv) root@netbox-lab:/opt/netbox/netbox# pip3 install netbox-topology-views
Collecting netbox-topology-views
  Downloading netbox_topology_views-3.1.1-py3-none-any.whl (358 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 358.7/358.7 kB 21.9 MB/s eta 0:00:00
Installing collected packages: netbox-topology-views
Successfully installed netbox-topology-views-3.1.1
(venv) root@netbox-lab:/opt/netbox/netbox# python3 manage.py migrate netbox_topology_views
Operations to perform:
  Apply all migrations: netbox_topology_views
Running migrations:
  Applying netbox_topology_views.0001_initial... OK
(venv) root@netbox-lab:/opt/netbox/netbox# python3 manage.py collectstatic --no-input

24 static files copied to '/opt/netbox/netbox/static', 520 unmodified.
(venv) root@netbox-lab:/opt/netbox/netbox# systemctl restart netbox netbox-rq
(venv) root@netbox-lab:/opt/netbox/netbox# pip list | grep netbox
netbox-topology-views         3.1.1
(venv) root@netbox-lab:/opt/netbox/netbox# grep ^VERSION /opt/netbox/netbox/netbox/settings.py
VERSION = '3.3.8'
(venv) root@netbox-lab:/opt/netbox/netbox# grep ^PLUGINS /opt/netbox/netbox/netbox/configuration.py
PLUGINS = ["netbox_topology_views"]
(venv) root@netbox-lab:/opt/netbox/netbox#
```

After these a Plugins menu appeared:

![image](https://user-images.githubusercontent.com/25136274/209426708-6f729caf-e61a-4c11-9287-3813217d70ac.png)

This was on Debian Linux 11.

**Reply by markkuleinio** [reply]
Also:

```
(venv) root@netbox-lab:/opt/netbox/netbox# python3 manage.py migrate netbox_topology_views
CommandError: No installed app with label 'netbox_topology_views'.
```

That is what I get when I **remove** this line from `configuration.py`:

```
PLUGINS = ["netbox_topology_views"]
```

**Comment by gechevara** [comment]
@treimers1892 add PLUGINS = ["netbox_topology_views"] to configuration.py at /opt/netbox/netbox/netbox , then cd /opt/netbox/netbox and go through installation sequence as usual.
